design/metadata_kv/index.md

```diff
@@ -0,0 +1,187 @@
+
+This is where concurrency control gets interesting, and where lakeFS is expected to provide a **correct** system whose semantics are well understood (lakeFS currently [falls short](https://github.com/treeverse/lakeFS/issues/2405) in that regard).
+
+Concurrency is more of an issue here because of how a commit works: when a commit starts, it scans the currently staged changes, applies them to the current commit pointed to by the branch, updating the branch reference and removing the staged changes it applied.
+
+Getting this right means we have to take care of the following:
+
+1. Ensure all staged changes that finished successfully before the commit started are applied as part of the commit (causality)
+1. Ensure acknowledged writes end up either in a resulting commit, or staged to be committed (no lost writes)
+
+To do this, we will employ 2 mechanisms:
+
+1. [Optimistic Concurrency Control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control) on the branch pointer using `SetIf()`
+1. Reliance on write [idempotency](https://en.wikipedia.org/wiki/Idempotence) provided by Graveler (i.e., Writing the same exact entry, with the same identity - will not appear as a change)
+
+This is what the proposed implementation will look like:
+
+#### Committer flow
+
+We add an additional field to each `Branch` object: In addition to the existing `staging_token`, we add an array of strings named `sealed_tokens`.
+
+1. get branch, find current `staging_token`
+1. use `SetIf()` to update the branch (if not modified by another process): push existing `staging_token` into `sealed_tokens`, set new uuid as `staging_token`.
+1. take the list of sealed tokens, and using the [`CombinedIterator()`](https://github.com/treeverse/lakeFS/blob/master/pkg/graveler/combined_iterator.go#L11), turn them into a single iterator to be applied on top of the existing commit
+1. Once the commit has been persisted (metaranges and ranges stored in object store, commit itself stored to KV using `Set()`), perform another `SetIf()` that updates the branch again: replacing its commit ID with the new value, and clearing `sealed_tokens`, as these have materialized into the new commit.
+1. If `SetIf()` fails, this means another commit is happenning/has happened on the same branch. Can either retry or let the caller know.
+
+An important property here is delegating safety vs liveness to a single optimistic `SetIf()` operation: if a commit fails somewhere along the process, a subsequent commit would simply pick up where the failed one left off,
+adding the current staging_token into the set of changes to apply. In an environment where there aren't many concurrent commits on the same branch, and that commits mostly succeed - the size of `sealed_tokens` would be relatively small. As an optimization, compaction strategies could be added to merge tokens together, but this *might not be necessary*, at least not for use cases we're familiar with.
```

Comment:
I am less optimistic (sorry) here.  Suppose that we have an automated commit, once every X seconds (X ~ 1s).  Typically everything works really well and we're happy.  I would claim that this is a good way to implement streaming input: write to branch `input`, every X seconds commit and on success merge to `main`.  (It is actually hard to do something else!  E.g. writing to branch `input-NNNNNN` with a running commit counter `NNNNNN` pushes this _huge_ problem onto the writers, who now need essentially to resolve the same issue that we are resolving here.)

Now suppose that our database (or maybe that part of our object store) has a hiccup and a commit takes >X seconds.  Then `sealed_tokens` starts growing.  AFAIU this requires every concurrent commit to fail (there is no point in retrying if `sealed_tokens` has grown since the last call, it will never go back to what we expect.  So typically _only the last concurrent commit_ will succeed.  So until some time after recovery, not only do we have no commits, but also _reads get progressively slower_.

It may be OK to leave this open in the first version(s), but I do want to document this limitation while we have it.

Reply:
Added a suggestion on how to address it. If you have other strategies that could be used here, let's add them!

Reply:
Thing is, we run into trouble _even if we **never** retry commits_ (so exponential backoff doesn't solve the issue)!  Every commit tries and fails, growing `sealed_tokens`.  As long as there's a nice fugue (commit starts by growing `sealed_tokens`, then runs until after the next commit starts) then no commit succeeds, and `sealed_tokens` grows, and reads take longer and longer.

I believe the only way around this will be to trim prefixes of `sealed_tokens`.  E.g. accept a commit if the end of its `sealed_tokens` overlaps that start of the current `sealed_tokens`, and trim the shared bit.

## Example

`sealed_tokens`: 1,2,3; `staging_token`: 4

Process A starts to commit tokens 1,2,3,4.

`sealed_tokens`: 1,2,3,4; `staging_token`: 5

Process B starts to commit tokens 1,2,3,4,5.

`sealed_tokens`: 1,2,3,4,5; `staging_token`: 6

Process C starts to commit tokens 1,2,3,4,5,6.

`sealed_tokens`: 1,2,3,4,5,6; `staging_token`: 7

Process A ends committing: it tries for a "fast-path commit", but since `staging_token` is no longer 5 that fails.  Instead it reads the branch (or uses the value return by Set-If, if we want to require such an interface).  Now 1,2,3,4 is a prefix of 1,2,3,4,5,6, so A tries again to commit the remaining `sealed_tokens` 5,6 and a `staging_token` of 7.  *That* works, so A succeeded!

`sealed_tokens`: 5,6; `staging_token`: 7

Process C ends committing: its fast-path commit works (`staging_token` is 7), so it drops all `sealed_tokens`.

`sealed_tokens`: (); `staging_token`: 7

Process B ends committing: its fast-path commit doesn't work, and it has an empty prefix, so it knows _it has already succeeded_.